### DISCUSSION TITLE
How to convert single-ended s4p parameters to s2p

### DISCUSSION THREAD

**Discussion by polcad** [scikit-rf/scikit-rf#950] [Q&A] [answered]
I am here  to ask for help writing a script that converts single ended s4p parameters file to an s2p parameters file.
I received an s4p file from a foundry and I cannot use it with my RF simulator -Qucs, because Qucs has no way to convert these files.
So I try to do the conversion with skrf. I read the file with no problem with the following code:
```python
import rf as rfpy
from pylab import *
import numpy as np
import re
import matplotlib
from matplotlib import pyplot as plt
import skrf as rf
#from skrf import Network, Frequency
rf.stylely

bfu_die = rf.Network('2020-E_combined2_DC_30GHz.s4p')
```
checking the result gives me:
```python
bfu_die
Out[119]: 4-Port Network: '2020-E_combined2_DC_30GHz',  0.0-30.0 GHz, 301 pts, z0=[50.+0.j 50.+0.j 50.+0.j 50.+0.j]
bfu_die.port_names
Out[122]: ['Chip_Net1', 'PCB_Net1', 'Chip_Net2', 'PCB_Net2']
```
I am not sure, but I think the port order is correct and I can run se2gmm(p=2) without having to reorder the ports:
```python
bfu_die.se2gmm(p=2)
bfu_die
Out[121]: 4-Port Network: '2020-E_combined2_DC_30GHz',  0.0-30.0 GHz, 301 pts, z0=[100.+0.j 100.+0.j  25.+0.j  25.+0.j]
bfu_die.port_names
Out[122]: ['Chip_Net1', 'PCB_Net1', 'Chip_Net2', 'PCB_Net2']
```
So what to do next and how to write the result to an s2p file that could be used directly with an RF simulator in 50 ohm system?

**Comment by cafeclimber** [comment]
If you know which ports you need to be a part of the network, you can extract the s parameters and create a new network setting the s parameters directly.

E.g.

```python
import numpy as np
import skrf

four_port = skrf.Network('file.s4p')
s = np.zeros((len(four_port.f), 2,2))
s[:,0,0] = four_port.s[:,1,1] # whichever port in the four port you want

# repeat for s[:,0,1], s[:,1,0], and s[:,1,1]

two_port = skrf.Network(frequency=four_port.frequency, s=s)

skrf.write('two_port.s2p', two_port)
```

There might be an easier way, but this should work

**Reply by polcad** [reply]
I don't think it is the right way to do it.
It is not just a matter of extracting the right two ports out of four like you suggest.
The s4p file contains single ended parameters which have to be converted to differential two port parameters.
The device is a die of an RF transistor. Plotting magnitude in dB of s4p gives all traces below zero. Transistor is an active device and if it was just a matter of selecting the right traces then some of them should be showing gain above 20dB.

**Reply by cafeclimber** [reply]
Ah sorry about that I missed what the goal was. I've never actually worked with mixed-mode parameters. 

I'm assuming you've looked [here](https://scikit-rf.readthedocs.io/en/latest/examples/mixedmodeanalysis/Mixed%20Mode%20S%20and%20Impedance%20Transformation.html)?

That tutorial describes a situation that sounds similar to yours, but taking a 3 port single-ended network into a two port network for use in a simulation software.

**Reply by polcad** [reply]
Yes, I saw the example from your link. The code from my first post is based on that example, but I do not know what to do next.
And I do not know how to write the right result to an s2p file usable in popular RF simulators.
I know that ADS has a dedicated tool that can do the conversion, but I do not have ADS.

**Reply by mhuser** [reply, marked as answer]
I think that the answer from @cafeclimber demonstrate how to [write](https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.network.Network.write_touchstone.html#skrf.network.Network.write_touchstone) the resulting network into an .s2p file.
Le'ts suppose this setup:
```
Original:
        +------+
CHIP1 0-|s0==s2|-2 CHIP2 
        |      | 
 PCB1 1-|s1==s3|-3 PCB2
        +------+
Renumbered for conversion:
        +------+                            +------+
CHIP1 0-|s0==s2|-2 PCB1           CHIP_dd 0-|d0==d1|-1 PCB_dd
        |      |        =se2gmm=>           |      |
CHIP2 1-|s1==s3|-3 PCB2           CHIP_cc 2-|c0==c1|-3 PCB_cc
        +------+                            +------+
```
Some port [renumbering](https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.network.Network.renumber.html#skrf.network.Network.renumber) seems to be required to achieve the port order required by [se2gmm](https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.network.Network.se2gmm.html#skrf.network.Network.se2gmm):
```Python
import skrf as rf
rf.stylely

# read
bfu_die = rf.Network('2020-E_combined2_DC_30GHz.s4p')
# renumber
bfu_die.renumber([0, 1, 2, 3], [0, 2, 1, 3])
# single-ended to mixed-modes
bfu_die.se2gmm(p=2)
# keep only differential, drop common mode.
bfu_die_dd = bfu_die.subnetwork([0, 1])
#write
bfu_die_dd.write_touchstone('bfu_die_dd.s2p')
```
If you want to simulate a perfect balun, you may want to force the ports of `bfu_die_dd` to 50 ohm. They will be 100 ohm otherwise.

**Comment by mattsoundworld** [comment]
Others have addressed how to get 2 ports from a 4 port network, so I'll focus on just the single ended to differential conversion. I did a cursory reading of se2gmm and found myself going in circles around the sub functions. I find the below algorithm to be more straightforward and intuitive, albeit less sophisticated and feature rich. 

However, it should be clear from below that your port assignment assumptions significantly impact the transform. Depending on the particular port setup of your device and measurement, you may need to play around with the swap_port call. A common convention with differential devices is one side evens and the other side odds ( 1 -> 2, 3 -> 4), eg, 1 and 3 form a coupled port, and 2 and 4 form a coupled port. But many transforms assume input ports are 1 and 2, and output ports are 3 and 4. Each approach is perfectly valid _if the the correct transform is applied!_

Due credit to TDLC for explaining the underlying math: 
[https://blog.teledynelecroy.com/2021/05/converting-single-to-mixed-mode-s.html](https://blog.teledynelecroy.com/2021/05/converting-single-to-mixed-mode-s.html)

I should note this assumes your reference impedance is 100 ohms differential, 25 ohms common... which can be a bad assumption for short length circuits which are not that. I cannot post the contents, but if this is an issue, download (ahem, buy) the below paper from IEEE and it will show you have to modify the M matrix below with appropriate keven and kodd coefficients. 

**Mixed-Mode S-Parameter Conversion for Networks with Coupled Differential Signals**
_Allan Huynh, Pär Håkansson and Shaofang Gong_

```
import skrf as rf
import math 
import numpy as np

def single_to_diff(S):
    #set up transform matrix
    M = np.array([[1,-1,0,0],[1,1,0,0],[0,0,1,-1],[0,0,1,1]])
    M = M * 1/math.sqrt(2)

    #swap ports so input ports are 1 and 2, output ports are 3 and 4
    S = swap_port(S, 2, 3)

    #iterate through and multiply M * S * inv(M)
    for x in range(len(S)):
        S[x, :, :] = np.matmul(np.matmul(M , S[x, :,:]), np.linalg.inv(M))

    return S

def swap_port(S, port1, port2):
    #add some sugar so we can reference ports 1 thru 4
    port1 = port1 - 1
    port2 = port2 - 1 
    
    #row swap
    cp = np.copy(S[:,port1,:])
    S[:, port1, :] = S[:, port2, :]
    S[:, port2,:] = cp

    #col swap
    cp = np.copy(S[:,:,port1])
    S[:, :, port1] = S[:, :, port2]
    S[:, :, port2] = cp
    return S

```
Usage would be as follows:

```
s4p = rf.Network(filepath)

Sdiff = single_to_diff(s4p.s)

```
Sdiff is still a 4x4 array with SDD, SDC, SCD, and SCC components. I don't recall off the top of my head which ports are the SDD ports, but if you got this far, you can handle that!
